### DISCUSSION TITLE
SensESP v3 alpha

### DISCUSSION THREAD

**Discussion by mairas** [SignalK/SensESP#676] [General]
Hey,

I have just pushed a new `dev-3` branch to the SensESP repo, representing an alpha version of SensESP v3. The new branch represents several months worth of development, mainly focusing in a complete rewrite of the web config UI. The config UI, written in React typescript, provides (hopefully) a much more polished experience, with user-friendly UIs for WiFi, SK and system settings. It also supports a plugin system, allowing creating application-specific config pages. WiFiManager is no longer used. Instead, the same UI is used both for initial WiFi setup and other system configuration.

Other noteworthy changes:

- Implemented HTTP Digest Authentication for the config UI, meaning that the UI is now (optionally) protected by a username and a password
- Got rid of the `Startable` base class because it was really not required.
- Use ArduinoJson 7
- Squash a bunch of bugs

Most of the internal plumbing is backwards-compatible. Some addons may require small changes to compile. For example, `ValueConsumer::set_input()` has been renamed to `::set()`.

The new code has had very little testing so far. If you're interested, please have a go and report the results. You can update existing projects by editing `platformio.ini` and replacing the SensESP dependency in the `lib_deps` section with the following:

    https://github.com/SignalK/SensESP.git#dev-3

**Comment by mestafin** [comment]
My initial comments on version 3. I converted an old project to the new version just to try it out.

I like the WiFi configuration, much more useful

I tried to clear the Signal settings and save the blank fields, but it reverts back to the old settings after a restart. 

Changing the hostname also reverts back to the old name after a restart

**Reply by mairas** [reply]
Have a gold medal and a fancy parrot for being first person to provide feedback on v3: 🏅, 🦜 

I assume you are referring to the Signal K settings? I have a nagging feeling I haven't implemented the changes in the backend yet. Should a quick fix.

**Reply by mairas** [reply]
Signal K settings should now be saved and restored properly. Please have a go.

**Comment by mestafin** [comment]
Thanks, working now

**Comment by mestafin** [comment]
Quick Question - how does the WiFi management now works?

I would like something like this:

Try Network 1
Try Network 2
Try Network 3
Activate AP for 2 minutes
Try Network 1
Try Network 2
......

This will allow the sensESP unit to connect to the available networks, even if it is not immediately available on boot.
The AP will allow you enough time to log in and configure any networks if not yet configured.

**Reply by mairas** [reply]
If you have both the AP and client configured, that's what it does - it keeps the AP on all the time while simultaneously connecting to another network as a client.

Frankly, I'm not sure how the WiFi channels behave - I don't think the AP and the STA (client) can have different channels, meaning that the AP channel would change whenever the device connects to another network as STA. But I haven't had disconnects during my testing. 🤷 

I don't think there's much downside having the AP on all the time.

**Reply by donmarlin** [reply]
I am playing with V3 right now. 

If you have the "client" connected to a server and then connect to the AP, the client will stop working. Once the AP is disconnected then you need to reboot the node for the client to connect again. I figure this is by design since having both up is probably not possible. I do like having AP available all the time.

**Comment by donmarlin** [comment]
For OTA is there any thought to having the OTA web interface?
Currently you can upgrade via ota thru platformIO or directly with the espota command line tool. 

The esp32 web ota is pretty handy though, especially for less technical users.
It might not be that hard since there is  already a webserver/menus etc running.

I might try to add myself but I have some learning to do first....:-)

**Reply by mairas** [reply]
Sounds like a nice idea, and yes, it would fit well in the new frontend. What would the use cases be, though?

**Reply by donmarlin** [reply]
The use case would be to allow "non-programmers" the ability to upgrade devices a little easier. Right now if someone has a node running Sensesp, the options to upgrade are: 1) using Arduino or PlatformIO or 2) using a command line tool like "espota" (comes with platformIO). This requires a fairly high level of knowledge or access to the tool chain. To me the value of Sensesp is once the application is written, the web UI is quite easy to use for "non programmers". This is especially true with V3. The new web UI is quite user friendly. If there was a web ui for upgrades it would provide the final piece of core functions. Just my 2 cents. 
Anyhow, Sensesp is already very feature rich and I know it is always tricky when to stop as far as releases go. I do like the challenge though so I will probably go playing myself. It would be a good way to learn how the web portion is structured. Thanks
